design/general-progress-monitoring.md

```diff
@@ -0,0 +1,496 @@
+
+    	
+    	// Progress  
+    	Progress(input *RestoreItemActionProgressInput) (OperationProgress, error)
+
+        // Cancel  
+    	Cancel(input *RestoreItemActionProgressInput) error
+    
+    // RestoreItemActionExecuteOutput contains the output variables for the ItemAction's Execution function.
+    type RestoreItemActionExecuteOutput struct {
+        // UpdatedItem is the item being restored mutated by ItemAction.
+        UpdatedItem runtime.Unstructured
+
+        // AdditionalItems is a list of additional related items that should
+        // be restored.
+        AdditionalItems []ResourceIdentifier
+
+        // SkipRestore tells velero to stop executing further actions
+        // on this item, and skip the restore step. When this field's
+        // value is true, AdditionalItems will be ignored.
+        SkipRestore bool
+
+        // OperationID is an identifier which indicates an ongoing asynchronous action which Velero will
+        // continue to monitor after restoring this item. If left blank, then there is no ongoing operation
+        OperationID string
+    }
+
+    // RestoreItemActionProgressInput contains the input parameters for the RestoreItemAction's Progress function.
+    type RestoreItemActionProgressInput struct {
+    	// Item is the item that was stored in the restore
+    	Item runtime.Unstructured
+    	// OperationID is the operation ID returned by RestoreItemAction Execute
+    	operationID string
+    	// Restore is the representation of the restore resource processed by Velero.
+    	Restore *velerov1api.Restore
+    }
+
+## Changes in Velero backup format
+
+No changes to the existing format are introduced by this change.  A
+`<backup-name>-itemoperations.json.gz` file will be added that contains the items and operation IDs
+(snapshotIDs) returned by VolumeSnapshotter and BackupItemAction plugins.  Also, the creation of the
+`velero-backup.json` object will not occur until the backup moves to one of the terminal phases
+(_Completed_, _PartiallyFailed_, or _Failed_).  Reconciliation should ignore backups that do not
+have a `velero-backup.json` object.
+
+The cluster that is creating the backup will have the Backup resource present and will be able to
+manage the backup before the backup completes.
+
+If the Backup resource is removed (e.g. Velero is uninstalled) before a backup completes and writes
+its `velero-backup.json` object, the other objects in the object store for the backup will be
+effectively orphaned.  This can currently happen but the current window is much smaller.
+
+### `<backup-name>-itemoperations.json.gz`
+The itemoperations file is similar to the existing `<backup-name>-itemsnapshots.json.gz`  Each snapshot taken via
+BackupItemAction will have a JSON record in the file.  Exact format TBD.
+
+## Changes to Velero restores
+
+A `<restore-name>-itemoperations.json.gz` file will be added that contains the items and operation IDs
+returned by RestoreItemActions.
```

Comment:
We are currently considering the design for the unified data path feature, that is, we put everything (so also include data from Velero's build-in data movers) into the backup repo (or kopia repo as it is the only candidate at present) through the unified repo interface, this is to support the scenarios in on-premise envs where there is no object store (i.e. only NFS is available). This means, we cannot assume that there is a sophisticated object store, then we have some restrictions for data persistency:

- All the data to be persisted must be in either a backup or a restore folder that is pre-created at the beginning of backup/restore
- For each backup/restore folder, the data to be persisted must be persisted by one single writer, otherwise, they must be written sequentially

Otherwise, we will have to develop something like a distributed file system where metadata is carefully taken care of to prevent mess up by in parallel write. That will be very complicated.

I am afraid the current itemoperation.json conflicts with the restriction if these json are written inside the BIA/RIA plugins, because we have no way to prevent them from concurrent write.

Of course, this BIA/RIA interfaces are for general purposes and not only for Velero build-in modules or build-in data movers. However, we should give a way for Velero build-in data movers, which are using these BIA/RIA interfaces, to delegate the json persistency to Velero to assure sequent write.

Alternatively, can we leave this json persistency thing to the specific plugin? If so, there will be no need to consider all the things above.

Reply:
@Lyndon-Li So these .json.gz files are just another file in the object store. We also have gzipped logs there per-backup/restore. If you can handle relocating the logs, then these would be handled the same way. In fact, for the use case of "on-permise NFS", I think it's probably more appropriate to handle that behind the object store abstraction. In other words, if we write an object store plugin that stores its data using on-premise NFS, then the velero backup/restore process won't care. It's just using `NewObjectStore` rather than `AwsObjectStore`.

Reply:
@sseago 
It is great that the plugin itself doesn't persist anything and the backup/restore controller does it instead. As the discussion, please check and add the clarification in the design to show that which part is done by the interface, which part is done by the specific plugin implementation and which one is done by velero. Thanks.

Reply:
@Lyndon-Li So the design already split out plugin API vs. backup/restore workflow changes. However, I noticed that the json files for ongoing operations was mentioned in both the workflow and plugin API sections, which may have contributed to the confusion. I've moved that content around so that it's not mentioned in the plugin API section now.